src/jit/importer.cpp

```diff
@@ -3281,32 +3281,84 @@ GenTreePtr Compiler::impInitializeArrayIntrinsic(CORINFO_SIG_INFO* sig)
+//       that is amenable to special downstream optimization opportunities
+//
+// Returns:
+//    IR tree to use in place of the call, or nullptr if the jit should treat
+//    the intrinsic call like a normal call.
+//
+//    pIntrinsicID set to non-illegal value if the call is recognized as a
+//    traditional jit intrinsic, even if the intrinsic is not expaned.
+//
+//    isSpecial set true if the expansion is subject to special
+//    optimizations later in the jit processing
+//
+// Notes:
+//    On success the IR tree may be a call to a different method or an inline
+//    sequence. If it is a call, then the intrinsic processing here is responsible
+//    for handling all the special cases, as upon return to impImportCall
+//    expanded intrinsics bypass most of the normal call processing.
+//
+//    Intrinsics are generally not recognized in minopts and debug codegen.
+//
+//    However, certain traditional intrinsics are identifed as "must expand"
+//    if there is no fallback implmentation to invoke; these must be handled
+//    in all codegen modes.
+//
+//    New style intrinsics (where the fallback implementation is in IL) are
+//    identified as "must expand" if they are invoked from within their
+//    own method bodies.
+//
+
+GenTree* Compiler::impIntrinsic(GenTree*                newobjThis,
+                                CORINFO_CLASS_HANDLE    clsHnd,
+                                CORINFO_METHOD_HANDLE   method,
+                                CORINFO_SIG_INFO*       sig,
+                                int                     memberRef,
+                                bool                    readonlyCall,
+                                bool                    tailCall,
+                                CORINFO_RESOLVED_TOKEN* pConstrainedResolvedToken,
+                                CORINFO_THIS_TRANSFORM  constraintCallThisTransform,
+                                bool                    isJitIntrinsic,
+                                CorInfoIntrinsics*      pIntrinsicID,
+                                bool*                   isSpecialIntrinsic)
 {
     bool              mustExpand  = false;
     bool              isSpecial   = false;
     CorInfoIntrinsics intrinsicID = info.compCompHnd->getIntrinsicID(method, &mustExpand);
     *pIntrinsicID                 = intrinsicID;
 
-    // Jit intrinsics are always optional to expand, and won't have an
-    // Intrinsic ID.
+    // Jit intrinsics won't have an IntrinsicID, and won't be identifed
+    // by the VM as must-expand.
     if (isJitIntrinsic)
     {
         assert(!mustExpand);
         assert(intrinsicID == CORINFO_INTRINSIC_Illegal);
+
+        // They however may still be must-expand. The convention we
+        // have adopted is that if we are compiling the intrinsic and
+        // it calls itself recursively, the recursive call is
+        // must-expand.
+        mustExpand = gtIsRecursiveCall(method);
```

Comment:
I wondered this recursive call checking may not enough for certain intrinsics in the future that have recursive functions as the fallback. I talked with @jkotas, he agreed to use `gtIsRecursiveCall` now, and we may need to keep this in mind and improve it later.